fifo/fifo_immutable_group.go

```diff
@@ -0,0 +1,139 @@
+package fifo
+
+import (
+	"github.com/google/martian/v3"
+	"github.com/google/martian/v3/verify"
+	"net/http"
+)
+
+// ImmutableGroup is a martian.RequestResponseModifier that maintains lists of
+// request and response modifiers executed on a first-in, first-out basis.
+// ImmutableGroup can be constructed only from a Group.
+// It cannot be modified.
+type ImmutableGroup struct {
+	reqmods         []martian.RequestModifier
+	resmods         []martian.ResponseModifier
+	aggregateErrors bool
+}
+
+// ModifyRequest modifies the request. By default, aggregateErrors is false; if an error is
+// returned by a RequestModifier the error is returned and no further modifiers are run. When
+// aggregateErrors is set to true, the errors returned by each modifier in the group are
+// aggregated.
+func (g *ImmutableGroup) ModifyRequest(req *http.Request) error {
+	var merr *martian.MultiError
+	for _, reqmod := range g.reqmods {
+		if err := reqmod.ModifyRequest(req); err != nil {
+			if g.aggregateErrors {
+				if merr == nil {
+					merr = martian.NewMultiError()
+				}
+				merr.Add(err)
+				continue
+			}
+
+			return err
+		}
+	}
+
+	if merr == nil || merr.Empty() {
+		return nil
+	}
+
+	return merr
+}
+
+// ModifyResponse modifies the request. By default, aggregateErrors is false; if an error is
+// returned by a RequestModifier the error is returned and no further modifiers are run. When
+// aggregateErrors is set to true, the errors returned by each modifier in the group are
+// aggregated.
+func (g *ImmutableGroup) ModifyResponse(res *http.Response) error {
+	var merr *martian.MultiError
+	for _, resmod := range g.resmods {
+		if err := resmod.ModifyResponse(res); err != nil {
+			if g.aggregateErrors {
+				if merr == nil {
+					merr = martian.NewMultiError()
+				}
+				merr.Add(err)
+				continue
+			}
+
+			return err
+		}
+	}
+
+	if merr == nil || merr.Empty() {
+		return nil
+	}
+
+	return merr
+}
+
+// VerifyRequests returns a MultiError containing all the
+// verification errors returned by request verifiers.
+func (g *ImmutableGroup) VerifyRequests() error {
```

Comment:
I not quite sure if ImmutableGroup should be able to verify req/res tho. Note that in the fifo.Group implementation this method uses WriteLock. 
Please double-check.